src/backend/distributed/transaction/worker_transaction.c

```diff
@@ -55,6 +56,59 @@ SendCommandToWorker(char *nodeName, int32 nodePort, char *command)
 }
 
 
+/*
+ * SendCommandToFirstWorker sends the given command only to the first worker node
+ * (sorted by group id) using SendCommandToWorker.
+ */
+void
+SendCommandToFirstWorker(char *command)
+{
+	List *workerNodeList = ActivePrimaryNodeList();
+	WorkerNode *firstWorkerNode = NULL;
+
+	workerNodeList = SortList(workerNodeList, CompareWorkerNodesByGroupId);
+	firstWorkerNode = (WorkerNode *) linitial(workerNodeList);
+
+	if (firstWorkerNode->groupId == GetLocalGroupId())
+	{
+		/*
+		 * Don't get shard resouce lock on the first worker, otherwise it
+		 * results with self-deadlock.
```

Comment:
This check seems out of place in `SendCommandToFirstWorker`. It's not what I would expect from the name of the function. Maybe we should do all of this in `LockShardListResourcesOnFirstWorker`.